File: Core/DriverHub.Application/Features/Entities/Reservations/Common/ReservationTimePolicy.cs
namespace DriverHub.Application.Features.Entities.Reservations.Common;

public static class ReservationTimePolicy
{
    public static int GetRentalDays(DateTime startDate, DateTime endDate) =>
        Math.Max(1, (int)Math.Ceiling((endDate - startDate).TotalDays));

    // Keep the Angular reservation-time-policy constant in sync.
    public const int StartGracePeriodMinutes = 5;

    public static bool IsStartAllowed(DateTime startDate) => IsStartAllowed(startDate, DateTime.UtcNow);

    public static bool IsStartAllowed(DateTime startDate, DateTime utcNow)
    {
        DateTime currentMinute = utcNow.AddTicks(-(utcNow.Ticks % TimeSpan.TicksPerMinute));
        return startDate >= currentMinute.AddMinutes(-StartGracePeriodMinutes);
    }
}

File: Client/DriverHub.Client/src/app/core/services/reservation/reservation-time-policy.ts
// Keep ReservationTimePolicy.StartGracePeriodMinutes in the API in sync.
export const RESERVATION_START_GRACE_PERIOD_MINUTES = 5;

export function minimumReservationStart(now = new Date()): Date {
  return new Date(Math.floor(now.getTime() / 60_000) * 60_000
    - RESERVATION_START_GRACE_PERIOD_MINUTES * 60_000);
}
